source/extensions/common/wasm/wasm_vm_base.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "extensions/common/wasm/wasm_vm.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+/**
+ * Wasm host stats.
+ */
+#define ALL_VM_STATS(COUNTER, GAUGE)                                                               \
+  COUNTER(created)                                                                                 \
+  COUNTER(cloned)                                                                                  \
+  GAUGE(active, NeverImport)
+
+struct VmStats {
+  ALL_VM_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+struct VmGlobalStats {
+  std::atomic<int64_t> active_;
+};
+
+// Wasm VM base instance. Provides common behavior (e.g. Stats).
+class WasmVmBase : public WasmVm {
+public:
+  WasmVmBase(Stats::ScopeSharedPtr scope, VmGlobalStats* global_stats_ptr,
+             absl::string_view runtime)
+      : scope_(scope), global_stats_ptr_(global_stats_ptr),
+        stats_(VmStats{
+            ALL_VM_STATS(POOL_COUNTER_PREFIX(*scope_, absl::StrCat("wasm_vm.", runtime, ".")),
```

Comment:
perf nit: capture this string concatenation in a member variable initialized earlier, otherwise it looks like it might be repeated in a macro a few times. Not sure if this would be material. Or you could make a static helper initialization method which captured the concatenation as a temp.